File: Semestrul 4/MPP/lab08/TriathlonNetworking/jsonprotocol/TriathlonClientJsonWorker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using log4net;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Triathlon.Model;
using Triathlon.Services;

namespace Triathlon.Network.JsonProtocol;

public class TriathlonClientJsonWorker : ITriathlonObserver, IDisposable
{
    private readonly ITriathlonServices _server;
    private readonly TcpClient _client;
    private readonly NetworkStream _stream;
    private readonly StreamReader _reader;
    private readonly StreamWriter _writer;
    private readonly JsonSerializer _jsonSerializer;
    private volatile bool _connected;

    private static readonly ILog _logger = LogManager.GetLogger(typeof(TriathlonClientJsonWorker));

    private static readonly Response _okResponse = JsonProtocolUtils.CreateOkResponse();

    public TriathlonClientJsonWorker(ITriathlonServices server, TcpClient client)
    {
        _server = server;
        _client = client;
        _jsonSerializer = new JsonSerializer();

        try
        {
            _stream = client.GetStream();
            _reader = new StreamReader(_stream);
            _writer = new StreamWriter(_stream) { AutoFlush = true };
            _connected = true;
            _logger.Info($"Worker initialized for client: {client.Client.RemoteEndPoint}");
        }
        catch (IOException ex)
        {
            _logger.Error("Error initializing worker", ex);
            _connected = false;
            throw;
        }
    }

    public void Run()
    {
        while (_connected)
        {
            try
            {
                                 if (!_client.Connected)
                {
                    _logger.Warn("Socket is closed, ending worker");
                    _connected = false;
                    break;
                }

                                 if (_client.Available > 0 && _reader.Peek() >= 0)
                {
                                         continue;
                }

                                 string requestLine = _reader.ReadLine();
                if (requestLine == null)
                {
                    _logger.Warn("Client disconnected (null read)");
                    _connected = false;
                    break;
                }

                Request request = JsonConvert.DeserializeObject<Request>(requestLine);
                Response response = HandleRequest(request);
                if (response != null)
                {
                    SendResponse(response);
                }
            }
            catch (SocketException se)
            {
                _logger.Error("Socket error, client likely disconnected", se);
                _connected = false;
                break;
            }
            catch (IOException e)
            {
                _logger.Error("IO error during request processing", e);
                _connected = false;
                break;
            }

                         try
            {
                Thread.Sleep(100);              }
            catch (ThreadInterruptedException e)
            {
                _logger.Warn("Worker thread interrupted", e);
                Thread.CurrentThread.Interrupt();              }
        }

                 CloseConnection();
    }

    private void CloseConnection()
    {
        try
        {
            _logger.Info($"Closing connection to client: {_client.Client.RemoteEndPoint}");

            if (_reader != null) _reader.Close();
            if (_writer != null) _writer.Close();
            if (_stream != null) _stream.Close();
            if (_client != null && _client.Connected) _client.Close();
        }
        catch (IOException e)
        {
            _logger.Error("Error closing connection", e);
        }
    }

    public void ResultsUpdated(Result result)
    {
        if (!_connected || !_client.Connected)
        {
            _logger.Warn("Cannot send update to disconnected client");
            throw new TriathlonException("Client is no longer connected");
        }

        Response resp = JsonProtocolUtils.CreateResultUpdatedResponse(result);
        _logger.Debug($"Sending result update to client {_client.Client.RemoteEndPoint} " +
                $"for participant: {result.Participant.Name} " +
                $"with points: {result.Points}");

        try
        {
            SendResponse(resp);
            _logger.Debug($"Successfully sent result update to client: {_client.Client.RemoteEndPoint}");
        }
        catch (IOException e)
        {
            _logger.Error($"Failed to send result update to client: {_client.Client.RemoteEndPoint}", e);
            _connected = false;
            throw new TriathlonException($"Sending error: {e}");
        }
    }

    private Response HandleRequest(Request request)
    {
        Response response = null;

        switch (request.Type)
        {
            case RequestType.LOGIN:
                _logger.Debug($"Login request...{request.Data}");
                var refereeJson = JsonConvert.SerializeObject(request.Data);
                Referee referee = JsonConvert.DeserializeObject<Referee>(refereeJson);
                try
                {
                    RefereeDiscipline assignment = _server.Login(referee, this);
                    return JsonProtocolUtils.CreateLoginResponse(assignment);
                }
                catch (TriathlonException e)
                {
                    _connected = false;
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }

            case RequestType.LOGOUT:
                _logger.Debug($"Logout request {request.Data}");
                var logoutRefereeJson = JsonConvert.SerializeObject(request.Data);
                Referee logoutReferee = JsonConvert.DeserializeObject<Referee>(logoutRefereeJson);
                try
                {
                    _server.Logout(logoutReferee, this);
                    _connected = false;
                    return _okResponse;
                }
                catch (TriathlonException e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }

            case RequestType.GET_RESULTS_BY_EVENT:
                _logger.Debug($"GetResultsByEvent request {request.Data}");
                var eventIdJson = JsonConvert.SerializeObject(request.Data);
                long eventId = JsonConvert.DeserializeObject<long>(eventIdJson);
                try
                {
                    List<Result> results = _server.GetResultsByEvent(eventId);
                    return JsonProtocolUtils.CreateGetResultsByEventResponse(results);
                }
                catch (Exception e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }

            case RequestType.UPDATE_RESULT:
                _logger.Debug($"UpdateResult request {request.Data}");
                var resultJson = JsonConvert.SerializeObject(request.Data);
                Result result = JsonConvert.DeserializeObject<Result>(resultJson);
                try
                {
                    _server.UpdateResult(result);
                    return _okResponse;
                }
                catch (Exception e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }

            case RequestType.GET_TOTAL_POINTS_FOR_PARTICIPANT:
                _logger.Debug($"GetTotalPointsForParticipant request {request.Data}");
                var paramsJson = JsonConvert.SerializeObject(request.Data);
                object[] parameters = JsonConvert.DeserializeObject<object[]>(paramsJson);

                                 long participantId = Convert.ToInt64(parameters[0]);
                long participantEventId = Convert.ToInt64(parameters[1]);

                try
                {
                    int points = _server.GetTotalPointsForParticipant(participantId, participantEventId);
                    response = JsonProtocolUtils.CreateOkResponse();
                    response.TotalPoints = points;
                    return response;
                }
                catch (Exception e)
                {
                    return JsonProtocolUtils.CreateErrorResponse(e.Message);
                }

            default:
                return null;
        }
    }

    private void SendResponse(Response response)
    {
                 if (!_client.Connected)
        {
            _logger.Warn("Attempted to send response on closed socket");
            throw new IOException("Connection is closed");
        }

        string responseLine = JsonConvert.SerializeObject(response);
        _logger.Debug($"Sending response: {responseLine}");

        lock (_writer)
        {
            _writer.WriteLine(responseLine);
            _writer.Flush();
        }
    }

    public void Dispose()
    {
        CloseConnection();
    }
}

File: Semestrul 4/MPP/lab11/java/TriathlonNetworking/src/main/java/triathlon/network/jsonprotocol/TriathlonClientJsonWorker.java
package triathlon.network.jsonprotocol;

import triathlon.model.Referee;
import triathlon.model.RefereeDiscipline;
import triathlon.model.Result;
import triathlon.services.ITriathlonObserver;
import triathlon.services.ITriathlonServices;
import triathlon.services.TriathlonException;
import com.google.gson.Gson;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.net.SocketException;
import java.util.List;

public class TriathlonClientJsonWorker implements Runnable, ITriathlonObserver {
    private ITriathlonServices server;
    private Socket connection;

    private BufferedReader input;
    private PrintWriter output;
    private Gson gsonFormatter;
    private volatile boolean connected;

    private static final Logger logger = LogManager.getLogger(TriathlonClientJsonWorker.class);

    public TriathlonClientJsonWorker(ITriathlonServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        gsonFormatter = new Gson();
        try {

            output = new PrintWriter(connection.getOutputStream(), true);
            input = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            connected = true;
            logger.info("Worker initialized for client: " + connection.getInetAddress());
        } catch (IOException e) {
            logger.error("Error initializing worker", e);
            connected = false;
        }
    }

    public void run() {
        while (connected) {
            try {

                if (connection.isClosed()) {
                    logger.warn("Socket is closed, ending worker");
                    connected = false;
                    break;
                }


                if (!input.ready()) {

                    continue;
                }


                String requestLine = input.readLine();
                if (requestLine == null) {
                    logger.warn("Client disconnected (null read)");
                    connected = false;
                    break;
                }

                Request request = gsonFormatter.fromJson(requestLine, Request.class);
                Response response = handleRequest(request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (SocketException se) {
                logger.error("Socket error, client likely disconnected", se);
                connected = false;
                break;
            } catch (IOException e) {
                logger.error("IO error during request processing", e);
                connected = false;
                break;
            }


            try {
                Thread.sleep(100);
            } catch (InterruptedException e) {
                logger.warn("Worker thread interrupted", e);
                Thread.currentThread().interrupt();
            }
        }


        closeConnection();
    }

    private void closeConnection() {
        try {
            logger.info("Closing connection to client: " + connection.getInetAddress());
            if (input != null) input.close();
            if (output != null) output.close();
            if (connection != null && !connection.isClosed()) connection.close();
        } catch (IOException e) {
            logger.error("Error closing connection", e);
        }
    }

    @Override
    public void resultsUpdated(Result result) throws TriathlonException {
        if (!connected || connection.isClosed()) {
            logger.warn("Cannot send update to disconnected client");
            throw new TriathlonException("Client is no longer connected");
        }

        Response resp = JsonProtocolUtils.createResultUpdatedResponse(result);
        logger.debug("Sending result update to client " + connection.getInetAddress() +
                " for participant: " + result.getParticipant().getName() +
                " with points: " + result.getPoints());

        try {
            sendResponse(resp);
            logger.debug("Successfully sent result update to client: " + connection.getInetAddress());
        } catch (IOException e) {
            logger.error("Failed to send result update to client: " + connection.getInetAddress(), e);
            connected = false;
            throw new TriathlonException("Sending error: " + e);
        }
    }

    private static final Response okResponse = JsonProtocolUtils.createOkResponse();

    private Response handleRequest(Request request) {
        Response response = null;

        if (request.getType() == RequestType.LOGIN) {
            logger.debug("Login request..." + request.getData());
            Referee referee = gsonFormatter.fromJson(gsonFormatter.toJson(request.getData()), Referee.class);
            try {
                RefereeDiscipline asssignement = server.login(referee, this);
                return JsonProtocolUtils.createLoginResponse(asssignement);
            } catch (TriathlonException e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }

        if (request.getType() == RequestType.LOGOUT) {
            logger.debug("Logout request " + request.getData());
            Referee referee = gsonFormatter.fromJson(gsonFormatter.toJson(request.getData()), Referee.class);
            try {
                server.logout(referee, this);
                connected = false;
                return okResponse;
            } catch (TriathlonException e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }

        if (request.getType() == RequestType.GET_RESULTS_BY_EVENT) {
            logger.debug("GetResultsByEvent request " + request.getData());
            Long eventId = gsonFormatter.fromJson(gsonFormatter.toJson(request.getData()), Long.class);
            try {
                List<Result> results = server.getResultsByEvent(eventId);
                return JsonProtocolUtils.createGetResultsByEventResponse(results);
            } catch (Exception e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }

        if (request.getType() == RequestType.UPDATE_RESULT) {
            logger.debug("UpdateResult request " + request.getData());
            Result result = gsonFormatter.fromJson(gsonFormatter.toJson(request.getData()), Result.class);
            try {
                server.updateResult(result);
                return okResponse;
            } catch (Exception e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }

        if (request.getType() == RequestType.GET_TOTAL_POINTS_FOR_PARTICIPANT) {
            logger.debug("GetTotalPointsForParticipant request " + request.getData());
            Object[] params = gsonFormatter.fromJson(gsonFormatter.toJson(request.getData()), Object[].class);
            Long participantId = ((Number) params[0]).longValue();
            Long eventId = ((Number) params[1]).longValue();
            try {
                int points = server.getTotalPointsForParticipant(participantId, eventId);
                response = JsonProtocolUtils.createOkResponse();
                response.setTotalPoints(points);
            } catch (Exception e) {
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }

        return response;
    }

    private void sendResponse(Response response) throws IOException {

        if (connection.isClosed()) {
            logger.warn("Attempted to send response on closed socket");
            throw new IOException("Connection is closed");
        }

        String responseLine = gsonFormatter.toJson(response);
        logger.debug("Sending response: " + responseLine);

        synchronized (output) {
            output.println(responseLine);
            output.flush();


            if (output.checkError()) {
                logger.error("PrintWriter encountered an error after write");
                throw new IOException("Error writing to socket output stream");
            }
        }
    }
}